### PR TITLE
PYTHON-4721 - Create individualized scripts for all shell.exec commands

### PR DESCRIPTION
**Pull request by NoahStapp**
[OCSP](https://spruce.mongodb.com/task/mongo_python_driver_ocsp_rhel8_latest_pypy3.10_test_ocsp_rsa_valid_cert_server_staples_patch_0733c4da44d4cd5c2a32c092762e639e14c3dc27_672cedc16512fc0008ef8421_24_11_07_16_43_10?execution=0&sortBy=STATUS&sortDir=ASC) and [AWS Auth](https://spruce.mongodb.com/task/mongo_python_driver_auth_aws_ubuntu_20_py3.13_aws_auth_test_latest_patch_0733c4da44d4cd5c2a32c092762e639e14c3dc27_672cedc16512fc0008ef8421_24_11_07_16_43_10?execution=0&sortBy=STATUS&sortDir=ASC) tests are still failing.

### CODE REVIEW

**Comment by blink1073** [comment]
We need to handle `SKIP_CSOT_TESTS` as well.

File: .evergreen/config.yml [hunk 1]
```diff
@@ -460,55 +326,67 @@ functions:
...
       params:
-        shell: "bash"
+        include_expansions_in_env: ["DRIVERS_TOOLS", "skip_EC2_auth_test"]
+        binary: bash
         working_dir: "src"
-        script: |
-          . .evergreen/scripts/env.sh
-          .evergreen/run-mongodb-aws-test.sh assume-role
+        args:
+          - .evergreen/scripts/run-mongodb-aws-test.sh assume-role
```
Comment (blink1073) [inline .evergreen/config.yml:347]: Do these need to get split up into two args as well?

File: .evergreen/config.yml [hunk 2]
```diff
@@ -532,110 +410,58 @@ functions:
...
-          . .evergreen/scripts/env.sh
-          if [ -f $DRIVERS_TOOLS/.evergreen/csfle/secrets-export.sh ]; then
-            . .evergreen/hatch.sh encryption:teardown
-          fi
-          rm -rf ${DRIVERS_TOOLS} || true
-          rm -f ./secrets-export.sh || true
+        args:
+          - .evergreen/scripts/cleanup.sh
 
   "fix absolute paths":
```
Comment (blink1073) [inline .evergreen/config.yml:430]: This, "windows fix",   "make files executable", and "init test-results" can be replaced with a call to https://github.com/mongodb-labs/drivers-evergreen-tools/blob/master/.evergreen/setup.sh

File: .evergreen/config.yml [hunk 3]
```diff
@@ -657,65 +483,55 @@ functions:
...
+        args:
+          - .evergreen/scripts/run-ocsp-test.sh
 
   "run-ocsp-server":
     - command: subprocess.exec
       params:
         background: true
         binary: bash
-        include_expansions_in_env: [SERVER_TYPE, OCSP_ALGORITHM]
+        include_expansions_in_env: ["SERVER_TYPE", "OCSP_ALGORITHM"]
```
Comment (blink1073) [inline .evergreen/config.yml:500]: We don't need quotes for these in general unless they have an expansion.

File: .evergreen/scripts/prepare-resources.sh [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+. src/.evergreen/scripts/env.sh
+set -o xtrace
+rm -rf $DRIVERS_TOOLS
+if [ "$PROJECT" = "drivers-tools" ]; then
+    # If this was a patch build, doing a fresh clone would not actually test the patch
+    cp -R $PROJECT_DIRECTORY/ $DRIVERS_TOOLS
+else
+    git clone https://github.com/mongodb-labs/drivers-evergreen-tools.git $DRIVERS_TOOLS
+fi
+echo "{ \"releases\": { \"default\": \"$MONGODB_BINARIES\" }}" >$MONGO_ORCHESTRATION_HOME/orchestration.config
```
Comment (blink1073) [inline .evergreen/scripts/prepare-resources.sh:12]: We can make the call to https://github.com/mongodb-labs/drivers-evergreen-tools/blob/master/.evergreen/setup.sh in this file.

File: .evergreen/scripts/run-ocsp-test.sh [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
```
Comment (blink1073) [inline .evergreen/scripts/run-ocsp-test.sh:2]: A general thought, what if we had a wrapper bash script that we run that invokes the env.sh file, uses `set -eu`, and then invokes the target script.

File: .evergreen/scripts/run-tests.sh [hunk 6]
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
+
+. .evergreen/scripts/env.sh
+if [ -n "${test_encryption}" ]; then
+  ./.evergreen/hatch.sh encryption:setup &
```
Comment (blink1073) [inline .evergreen/scripts/run-tests.sh:5]: We might need to break this out the way we handle it in the go driver, as a separate background command: https://github.com/mongodb/mongo-go-driver/blob/cf0348c9a63f6e4086e2c62aa62e70a9d3b3a986/.evergreen/config.yml#L508

File: .evergreen/scripts/teardown-aws.sh [hunk 7]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
```
Comment (blink1073) [inline .evergreen/scripts/teardown-aws.sh:1]: We can merge some of these teardown scripts into `cleanup.sh`.

File: .evergreen/config.yml [hunk 8]
```diff
@@ -288,167 +239,82 @@ functions:
...
-            COMPRESSORS=${COMPRESSORS} \
-            AUTH=${AUTH} \
-            SSL=${SSL} \
-            TEST_DATA_LAKE=${TEST_DATA_LAKE} \
-            TEST_SUITES=${TEST_SUITES} \
-            MONGODB_API_VERSION=${MONGODB_API_VERSION} \
-            SKIP_HATCH=${SKIP_HATCH} \
-            bash ${PROJECT_DIRECTORY}/.evergreen/hatch.sh test:test-eg
+        binary: bash
+        include_expansions_in_env: ["AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY", "AWS_SESSION_TOKEN"]
```
Comment (blink1073) [inline .evergreen/config.yml:295]: Note you'll need all of the expansions used above, or to put them in the env.sh file.

Comment (NoahStapp) [reply .evergreen/config.yml:295]: I prefer putting as many as possible in the env.sh file to reduce the amount of user error when adding new tasks.

File: .evergreen/scripts/run-ocsp-test.sh (hunk 5 above)
Comment (NoahStapp) [reply .evergreen/scripts/run-ocsp-test.sh:2]: I like that idea. More standardization is always good.

File: .evergreen/scripts/teardown-aws.sh (hunk 7 above)
Comment (blink1073) [reply .evergreen/scripts/teardown-aws.sh:1]: Yeah, note how we handle it in the drivers-tools teardown: https://github.com/mongodb-labs/drivers-evergreen-tools/blob/aeb2d4af763ca5eab3bb29acb8038042d4268740/.evergreen/teardown.sh#L15

File: .evergreen/config.yml (hunk 8 above)
Comment (blink1073) [reply .evergreen/config.yml:295]: Yeah that sounds right to me.

File: .evergreen/config.yml (hunk 2 above)
Comment (NoahStapp) [reply .evergreen/config.yml:526]: This change should take place in a separate PR. There are enough changes already for a single PR.

File: .evergreen/scripts/configure-env.sh [hunk 9]
```diff
@@ -54,15 +54,8 @@ export MONGODB_API_VERSION="${MONGODB_API_VERSION:-}"
 export SKIP_HATCH="${SKIP_HATCH:-}"
 export skip_crypt_shared="${skip_crypt_shared:-}"
 export STORAGE_ENGINE="${STORAGE_ENGINE:-}"
-export ORCHESTRATION_FILE="${ORCHESTRATION_FILE:-}"
 export REQUIRE_API_VERSION="${REQUIRE_API_VERSION:-}"
 export LOAD_BALANCER="${LOAD_BALANCER:-}"
```
Comment (blink1073) [inline .evergreen/scripts/configure-env.sh:58]: LOAD_BALANCER is set at the task level

File: .evergreen/config.yml (hunk 2 above)
Comment (blink1073) [reply .evergreen/config.yml:526]: I created https://jira.mongodb.org/browse/PYTHON-4965 to track this.

**Review by blink1073 — APPROVED** [review]
LGTM, thanks!  I scheduled several other variants to make sure all is well.

**Comment by blink1073** [comment]
It looks like "upload mo artifacts" is causing teardown failures.

**Comment by blink1073** [comment]
The encryption servers still need handling: ` [2024/11/12 14:07:25.482] Could not detect PID file`

**Comment by NoahStapp** [comment]
> The encryption servers still need handling: ` [2024/11/12 14:07:25.482] Could not detect PID file`

Looking into this one, it's oddly inconsistent.

**Comment by blink1073** [comment]
FYI I added `Encryption RHEL8 Python3.10` to the default tasks that run.

**Comment by blink1073** [comment]
I scheduled a cross-section of build hosts, there were a few failures

**Comment by blink1073** [comment]
It looks like we can't use PYTHON_BINARY yet for `run-orchestration`: 

```
[2024/11/19 11:07:18.874]   File "/Library/Frameworks/Python.Framework/Versions/3.13/lib/python3.13/urllib/request.py", line 1322, in do_open
[2024/11/19 11:07:18.874]     raise URLError(err)
[2024/11/19 11:07:18.874] urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1020)>
```

Something might be off about the 3.13 toolchain because of the free-threading build.

**Comment by NoahStapp** [comment]
> It looks like we can't use PYTHON_BINARY yet for `run-orchestration`:
> 
> ```
> [2024/11/19 11:07:18.874]   File "/Library/Frameworks/Python.Framework/Versions/3.13/lib/python3.13/urllib/request.py", line 1322, in do_open
> [2024/11/19 11:07:18.874]     raise URLError(err)
> [2024/11/19 11:07:18.874] urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1020)>
> ```
> 
> Something might be off about the 3.13 toolchain because of the free-threading build.

Correct, I went ahead and removed `PYTHON_BINARY` from anywhere it wasn't before.

**Comment by blink1073** [comment]
Hold off on further pushes for a bit please, I scheduled a cross-section of tasks.  So far we have:

no-server and other-hosts tests are failing with: `.evergreen/scripts/run-tests.sh: line 44: PYTHON_BINARY: unbound variable`.

**Comment by blink1073** [comment]
looks like serverless is missing auth: `OperationFailure('(Unauthorized) not authorized on admin to execute command`

**Comment by blink1073** [comment]
AWS failure: `.evergreen/scripts/run-with-env.sh: line 21: AWS_ROLE_SESSION_NAME="test": No such file or directory`

**Comment by blink1073** [comment]
Encryption PyOpenSSL RHEL8 PyPy3.10 - maybe missing SSL config: `AssertionError: "IP address mismatch|wronghost|IPAddressMismatch|Certificate" does not match "SSL handshake failed`

**Comment by NoahStapp** [comment]
> looks like serverless is missing auth: `OperationFailure('(Unauthorized) not authorized on admin to execute command`

It's running with auth, didn't we see this error previously as well?

**Comment by blink1073** [comment]
Ah, yep, [PYTHON-4963](https://jira.mongodb.org/browse/PYTHON-4963).

**Review by blink1073 — APPROVED** [review]
LGTM, thank you!

**Pull request merged**